lib/browser/api/web-contents.ts

```diff
@@ -553,43 +577,60 @@ WebContents.prototype._init = function () {
   if (this.getType() !== 'remote') {
     // Make new windows requested by links behave like "window.open".
     this.on('-new-window' as any, (event: any, url: string, frameName: string, disposition: string,
-      rawFeatures: string, referrer: string, postData: string) => {
-      const { options, webPreferences, additionalFeatures } = parseFeatures(rawFeatures);
-      const mergedOptions = {
-        show: true,
-        width: 800,
-        height: 600,
-        title: frameName,
-        webPreferences,
-        ...options
-      };
+      rawFeatures: string, referrer: any, postData: any) => {
+      openGuestWindow({
+        event,
+        embedder: event.sender,
+        disposition,
+        referrer,
+        postData,
+        overrideBrowserWindowOptions: {},
+        windowOpenArgs: {
+          url,
+          frameName,
+          features: rawFeatures
+        }
+      });
+    });
 
-      internalWindowOpen(event, url, referrer, frameName, disposition, mergedOptions, additionalFeatures, postData);
+    let windowOpenOverriddenOptions: BrowserWindowConstructorOptions | null = null;
```

Comment:
Can we guarantee that there is no possible race condition here?  I'd be concerned that enough `window.open` calls could result in us assigning the wrong `webPreferences` to the new window

Reply:
Yes it is guaranteed, since both the `-will-add-new-contents` and `-add-new-contents` are fired sequentially from the blue box in the diagram. Specifically, they are both fired from the same method in Chrome: `WebContents::CreateNewWindow`, in the delegate callbacks for `delegate_->IsWebContentsCreationOverridden` and `delegate_->AddNewContents`.

Reply:
You're right, it needs to be safer. Hmm. Back to the drawing board 🤔

Reply:
Upon further review, it actually is safe. The reason it's safe is because the params are associated with the new WebContents in `WebContentsCreatedWithFullParams` (our patched version of `WebContentsCreated`) via the `WebContentsPreferences` which associates the WebPreferences with the WebContents by `content::WebContentsUserData`. This is called before the AddNewContents paths branch. Obviously this is not clear enough because we both missed it and I wrote the code 😂 😭  So I'll add some comments for clarity.

Reply:
Is there some way we can make this more direct in the code? It feels very fragile and prone to breaking in obscure situations, especially if the upstream code changes at all.

Reply:
I mean, we could always patch the hell out of Chrome, but I was really trying to avoid that. 

If I could find a more direct way, I definitely would've written it that way 😄 Not to say there isn't one, but it is apparently beyond my ability. I welcome suggestions though!

Reply:
It looks to me like `ShowCreatedWindow` isn't guaranteed to be called after `IsWebContentsCreationOverridden`, not immediately, and not necessarily _ever_.

Here's the way that can happen:
1. Renderer calls [CreateNewWindow](https://source.chromium.org/chromium/chromium/src/+/master:content/renderer/render_view_impl.cc;l=440;drc=04ccda060610036333f90b365e21fa76bbc684b9)
2. Renderer creates a OnceCallback that will call `ShowCreatedWindow`. It passes the callback to...
3. [RenderView::Create](https://source.chromium.org/chromium/chromium/src/+/master:content/renderer/render_view_impl.cc;l=519;drc=04ccda060610036333f90b365e21fa76bbc684b9) (this is guaranteed to happen in all cases). RenderView::Create calls...
4. [RenderView::Initialize](https://source.chromium.org/chromium/chromium/src/+/master:content/renderer/render_view_impl.cc;l=327;drc=929e790ad17a54008b3dac86cbe30751c968c2ce).
5. If `!local_main_frame`, the `ShowCreatedWindow` callback is dropped, and the callback will never be invoked.

I'm not 100% sure of all the cases in which this can occur, but possibly when opening cross-origin windows?

Separately, in the case that `local_main_frame` is true, the callback isn't invoked immediately. Instead, it's [saved](https://source.chromium.org/chromium/chromium/src/+/master:content/renderer/render_widget.cc;l=260;drc=929e790ad17a54008b3dac86cbe30751c968c2ce) in RenderWidget::Initialize, and only later invoked in [RenderWidget::Show](https://source.chromium.org/chromium/chromium/src/+/master:content/renderer/render_widget.cc;l=399;drc=929e790ad17a54008b3dac86cbe30751c968c2ce). The latter is not guaranteed to be called, in particular [if an old window was reused](https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/core/page/create_window.cc;l=316;drc=929e790ad17a54008b3dac86cbe30751c968c2ce).

I think the way to trigger this case would be to try to `window.open` twice with the same frame name.

Reply:
I tried opening multiple windows with the same frame name, and also opening multiple cross-origin child windows, and I wasn't able to observe any interleaving.

Reply:
I'm going to resolve this for now. I'd like to add a test for this but I don't really know what to test for. I'm not able to observe any misbehavior currently, and I'm not sure how to anticipate what might trigger misbehavior in the future. Let's be on the lookout for failures in this area in future and be prepared to respond + refactor.